lightningd/subd.c

```diff
@@ -191,65 +210,32 @@ static int subd(const char *path, const char *name,
 		goto close_execfail_fail;
 
 	if (childpid == 0) {
-		int fdnum = 3, stdin_is_now = STDIN_FILENO;
 		size_t num_args;
 		char *args[] = { NULL, NULL, NULL, NULL, NULL };
+		int *fds = tal_arr(tmpctx, int, 1);
 
 		close(childmsg[0]);
 		close(execfail[0]);
 
-		// msg = STDIN
-		if (childmsg[1] != STDIN_FILENO) {
-			/* Do we need to move STDIN out the way? */
-			stdin_is_now = dup(STDIN_FILENO);
-			if (!move_fd(childmsg[1], STDIN_FILENO))
-				goto child_errno_fail;
-		}
-
-		/* Dup any extra fds up first. */
+		/* msg = STDIN */
+		fds[0] = childmsg[1];
 		while ((fd = va_arg(*ap, int *)) != NULL) {
-			int actual_fd = *fd;
-			/* If this were stdin, we moved it above! */
-			if (actual_fd == STDIN_FILENO)
-				actual_fd = stdin_is_now;
-
-			/* If we would overwrite important fds, move those.  */
-			if (fdnum == dev_disconnect_fd) {
-				if (!move_fd_any(&dev_disconnect_fd))
-					goto child_errno_fail;
-			}
-			if (fdnum == execfail[1]) {
-				if (!move_fd_any(&execfail[1]))
-					goto child_errno_fail;
-			}
-
-			if (!move_fd(actual_fd, fdnum))
-				goto child_errno_fail;
-			fdnum++;
+			assert(*fd != -1);
+			tal_arr_expand(&fds, *fd);
 		}
 
-		/* Move dev_disconnect_fd *after* the extra fds above.  */
-		if (dev_disconnect_fd != -1) {
-			/* Do not overwrite execfail[1].  */
-			if (fdnum == execfail[1]) {
-				if (!move_fd_any(&execfail[1]))
-					goto child_errno_fail;
-			}
-			if (!move_fd(dev_disconnect_fd, fdnum))
-				goto child_errno_fail;
-			dev_disconnect_fd = fdnum;
-			fdnum++;
-		}
+		/* If we have a dev_disconnect_fd, add it after. */
+		if (dev_disconnect_fd != -1)
+			tal_arr_expand(&fds, dev_disconnect_fd);
+
+		/* Finally, the fd to report exec errors on */
+		tal_arr_expand(&fds, execfail[1]);
 
-		/* Move execfail[1] *after* the fds we will pass
-		 * to the subdaemon.  */
-		if (!move_fd(execfail[1], fdnum))
+		if (!shuffle_fds(fds, tal_count(fds)))
```

Comment:
This changes the `fd` number of `execfail[i]` and `dev_disconnect_fd`, but does not seem to update those variables?  Maybe `shuffle_fds` should accept an array of `int *` instead of an array of `int`?